src/backend/distributed/planner/relation_restriction_equivalence.c

```diff
@@ -1052,3 +1250,104 @@ AttributeEquivalancesAreEqual(AttributeEquivalenceClass *firstAttributeEquivalan
 
 	return true;
 }
+
+
+/*
+ * ContainsUnionSubquery gets a queryTree and returns true if the query
+ * contains
+ *      - a subquery with UNION set operation
+ *      - no joins above the UNION set operation in the query tree
+ *
+ * Note that the function allows top level unions being wrapped into aggregations
+ * queries and/or simple projection queries that only selects some fields from
+ * the lower level queries.
+ *
+ * If there exists joins before the set operations, the function returns false.
+ * Similarly, if the query does not contain any union set operations, the
+ * function returns false.
+ */
+bool
+ContainsUnionSubquery(Query *queryTree)
+{
+	List *rangeTableList = queryTree->rtable;
+	Node *setOperations = queryTree->setOperations;
+	List *joinTreeTableIndexList = NIL;
+	Index subqueryRteIndex = 0;
+	uint32 joiningRangeTableCount = 0;
+	RangeTblEntry *rangeTableEntry = NULL;
+	Query *subqueryTree = NULL;
+
+	ExtractRangeTableIndexWalker((Node *) queryTree->jointree, &joinTreeTableIndexList);
+	joiningRangeTableCount = list_length(joinTreeTableIndexList);
+
+	/* don't allow joins on top of unions */
+	if (joiningRangeTableCount > 1)
+	{
+		return false;
+	}
+
+	subqueryRteIndex = linitial_int(joinTreeTableIndexList);
+	rangeTableEntry = rt_fetch(subqueryRteIndex, rangeTableList);
+	if (rangeTableEntry->rtekind != RTE_SUBQUERY)
+	{
+		return false;
+	}
+
+	subqueryTree = rangeTableEntry->subquery;
+	setOperations = subqueryTree->setOperations;
+	if (setOperations != NULL)
+	{
+		SetOperationStmt *setOperationStatement = (SetOperationStmt *) setOperations;
+
+		if (setOperationStatement->op != SETOP_UNION)
```

Comment:
may be a single line comment here stating set operation type in whole tree is checked elsewhere